src/sample.py

```diff
@@ -39,7 +40,7 @@ def message (self, buf, font, msg, data, _x_what_is_this):
 		return True
 debugger = Debugger()
 hb.buffer_set_message_func (buf, debugger.message, 1, 0)
-hb.buffer_add_utf8 (buf, text.encode('utf-8'), 0, -1)
+hb.buffer_add_codepoints (buf, codepoints, 0, -1)
```

Comment:
`hb_buffer_add_codepoints()` does not validate the text and i’m not sure if Python strings can never contain invalid Unicode characters, it is better to use the `_utf32` method as it will work with the same input but HarfBuzz will also make sure the input is valid.